changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml

```
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: sdk/nodejs
+  description: Support loading package.json from parent directory. If `package.json` is not found in cwd, Pulumi recursively searches up the directory tree until it is found. If `package.json` provides a `main` field, that field is relative to cwd.
```

wording suggestions:

```suggestion
  description: Support loading package.json from parent directory. If `package.json` is not found in the current directory, Pulumi recursively searches up the directory tree until it is found. If `package.json` provides a `main` field, that field is relative to the current directory.
```

---

More importantly:

> If `package.json` provides a `main` field, that field is relative to the current directory

Isn't that supposed to be relative to the directory containing package.json, not the current directory?
Is the behavior incorrect or the description?

---

> Isn't that supposed to be relative to the directory containing package.json, not the current directory?

So I thought about this a bit when writing the test cases, and I believe the answer is "no", but I definitely want to hear your thoughts.

The most common use-case for putting `package.json` above pwd is a monorepo, where that `package.json` is serving multiple Pulumi projects. Suppose that `package.json` has a `main` field. That field will always point to the same entrypoint, no matter which Pulumi project is running. 

Instead, if `main` is relative to `pwd`, then you can universally rename your entrypoint. For example, suppose you wanted your entrypoint to be `main.ts` in every Pulumi project. But providing `"main": "main.ts"` in your `package.json`, you will override the default `index.js` for all Pulumi projects in the monorepo. I think that's the only time this would make sense.

----------

This is what you'd have if `main` is relative to `pwd`.
```
.
└── my-monorepo/
    ├── package.json
    ├── node_modoules/
    │   ├── @pulumi/pulumi
    │   └── ...
    ├── project-A/
    │   ├── Pulumi.yaml
    │   └── main.ts
    └── project-B/
        ├── Pulumi.yaml
        └── main.ts
```

This is what you'd have if main is relative to `package.json`
```
.
└── my-monorepo/
    ├── package.json
    ├── main.ts
    ├── node_modoules/
    │   ├── @pulumi/pulumi
    │   └── ...
    ├── project-A/
    │   └── Pulumi.yaml
    └── project-B/
        └── Pulumi.yaml
```

This precludes monorepos from using this feature meaningfully.

---

Thanks for the `@`, @RobbieMcKinstry otherwise I would have missed this.
I see what you're trying to do with the UX here, but I don't think we have that flexibility here.
Per the [NPM docs](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#main)

> This should be a module ID relative to the root of your package folder.

We cannot change that behavior—users expect main to be relative to the package directory—the one containing the package.json

But you're right, this exposes a big gap in this feature.
Moving the package.json up one level still won't allow users to have multiple Pulumi programs coexist in the same NPM package, or even a normal program and a Pulumi program in the same package.
We need to find a different dial for that, though. As mentioned above, I don't think we can change the meaning of what `"main"` means in the package.json.
One option I can imagine is a field for the JS backend in the Pulumi.yaml that specifies the script to run, defaulting to the `main` specified in the package.json. That value can be relative to the Pulumi.yaml directory.

---

Ok, I'll get this to work using main relative to package.json instead of pwd to adhere to the spec. :)

This feature is still useful even if main has to be relative to package.json. This will be an improvement for anyone in a monorepo using index.js or index.ts as an entry point, which is the default across all Node projects, in addition to the subset which is Pulumi Node projects.

---

@abhinav I've updated the tests to reflect the `npm` standard behavior. This should be more than sufficient. I've learned that if you're using a monorepo, you're probably _**not**_ using `main`, because that would defeat the purpose.